### ISSUE TITLE
Stand alone projects located under another project fails to load

### ISSUE THREAD

**Issue by eskatos** [kelemen/netbeans-gradle-project#241]
When opening a project whose directory tree contains another standalone gradle project the latter is automatically opened and fails to.

It is very common to have such a setup on gradle plugins for integration testing purpose.

An example project is the [Gradle](https://github.com/gradle/gradle) source code itself. It has ~60 standalone projects in integration tests resources directories so it eats up a lot of time/cpu/memory to simply attempt to load them.
Many gradle plugins also have such a setup for integration testing purpose.

I first raised this problem in #212. This issue is to distinguish it from #212 that is actually different.

Here is what @kelemen said about it, quoting for reference;

> @eskatos What you are seeing is a different issue: The problem is that those project are supposed to be stand alone (and are not part of the Gradle project). However Gradle - by default - will believe them to be part of the project because it finds the settings.gradle up in the directory hierarchy. The only way I can think of to solve the problem is to check for the error message and try reloading the project given no settings.gradle (I'm not sure if I can tell Gradle that there is no settings.gradle but I guess it should be possible).

**Comment by eskatos** [comment]
From my point of view theses 'included standalone projects' should not be loaded at all when opening the overall project. Gradle doc for [multi-projects builds](https://docs.gradle.org/current/userguide/build_lifecycle.html#sec:multi_project_builds) state that 'subprojects' must be declared in `settings.gradle`. In the example builds mentioned above, the incriminated projects are clearly not registered in `settings.gradle`. Or I missed something?

**Comment by kelemen** [comment]
Sadly, I can't really not load these projects because they are automatically requested to be loaded by NetBeans and I have no way to determine why it is being loaded (it might be the case that they are really needed). Besides, from the plugin's point of view they are simply (sub)projects. I workaround this issue by zip-ing these kind of Gradle projects (to avoid automatic project loading).

They are not listed in the settings.gradle, that is true. However, Gradle has to determine somehow where the root project is given a particular subproject (by directory). This is documented in [Chapter 58](https://docs.gradle.org/current/userguide/build_lifecycle.html#sec:initialization).

**Comment by eskatos** [comment]
I see. Is it the same "*no way to determine why it is being loaded*" that was at play when you fixed a project loading issue sometime in the past where simply giving focus to project nodes in open dialogs triggered projects loading? Any chance that somethings get fixed/added in Netbeans about that? Some issue to report/vote on?

**Comment by eskatos** [comment]
In my previous comments I was referring to #25.

**Comment by kelemen** [comment]
I make exceptions on a case-by-case basis. That is, I try to guess that in some cases the info NetBeans is asking the plugin is not that important in which case I don't load the project. But the reason for this project to be loaded is that NetBeans is looking for the classpath of the sources which I cannot ignore (since it does the same thing for valid reasons too).

**Comment by eskatos** [comment]
Thanks for your feedback @kelemen.

So, if I understand correctly this boils down to Netbeans not providing enough context to filter out the loading request. Is that so?

On the other hand, isn't it possible to say something like "netbeans ask to load this particular project located inside that other loaded project but it is not part of it's submodules defined in `settings.gradle` so we could skip it" ?

**Comment by JustGregory** [comment]
Then why _ever_ have within ``settings.gradle`` an iterator to look in subdirectories for Gradle projects at all?? It's the *ITERATOR* that is looking through subdirectories and including subprojects, which is why the above would happen. You don't want that? Take out the iterator, but manually specify each and every subproject with an ``include`` command. It's doable.... I do it.

In my case, in EVERY Gradle main project, I'm editing the build.gradle, settings.gradle, and common.gradle for my own personal style, anyway. In the case of settings, though, I believe the project-creation dialog could be enhanced to allow entering into a list box, an initial listing of subprojects (one per line) to automatically create after creating the main project. That way, when creating a new main project, we are starting with a manual listing of project includes, instead of ... iterators. Or, the wizard dialog could allow this as an alternative, which I would prefer to use.

**Comment by kelemen** [comment]
@eskatos The poblem with that is that at the time the project is being loaded, the root project is not necessarily already loaded. Though I could first force loading the root project first and then decide if this project is part of that. I will think about this a little. As for "loading request": I just call it so, NetBeans does not actually does that but rather, starts asking questins like "what is the classpath of this file/directory" which forces me to evaluate the scripts.

@JustGregory I assume you are talking about the wizard. Eskatos is not talking about wizard generated projects and does not have a problem with the build scripts but rather, with the plugin's assumptions (the script is fine). But to answer your question: The wizard was done so because it requires less understanding from the user this way (I don't want the plugin to edit the `settings.gradle` because it is a code). By the way, if you already familiar with Gradle, you probably don't want to use the wizard anyway (except for very small projects). I personally prefer creating Gradle tasks to generate subprojects.

**Comment by JustGregory** [comment]
Actually, @kelemen, you partially misunderstand my comment. I am talking about the inclusion iterator in a default new ``settings.gradle`` as the beginning of my comment, and mention the wizard in the latter half of it. The mention of the wizard dialog is my attempt to suggest a replacement for an iterator that causes least pain. (I see now that it might be better for me to create a new "enhancement" issue, so I did. See #242 for more detail on my thoughts.)

**Comment by eskatos** [comment]
> @kelemen 
> Though I could first force loading the root project first and then decide if this project is part of that. I will think about this a little.

Did you find some time to think about it?

**Comment by kelemen** [comment]
It is not trivial to implement (relatively cleanly) but currently I consider this the second issue to be fixed after #195. I cannot exactly tell you though when I will implement it. Although, my problems with fixing this is mostly academical :).

**Comment by eskatos** [comment]
Sounds good :-)

**Comment by kelemen** [comment]
This should be fixed in master. I plan to have a new release soon (i.e., this week or first week of next year).

**Comment by eskatos** [comment]
Muuch better! Thank you and happy new year :tada:
